### PR TITLE
Simplify composition handling

### PR DESCRIPTION
**Pull request by treeowl**
Theme: follow the structure of the types instead of tearing
them up and reconstituting them.

### CODE REVIEW

**Comment by treeowl** [comment]
@RyanGlScott, I pulled some of the ideas out of my other PR that I think make the regular mechanism a little less convoluted. Might also avoid some of the repeated checks, but I'm not sure about that.

**Comment by treeowl** [comment]
In the (common) case where someone uses a `deriveAll` form, certain checks in the `from1` and `to1` deriving are redundant. Worth fixing? Unsure.

**Comment by treeowl** [comment]
Hmm.... The second commit will reduce performance somewhat, I imagine, because it performs the recursive dusting in `Q`. Maybe I shouldn't make it check for higher-rank types?

File: src/Generics/Deriving/TH/Internal.hs [hunk 1]
```diff
@@ -145,52 +145,65 @@ makeFunKind = makeFunType
...
+    check d tcName = do
+      mbinders <- getTypeFamilyBinders tcName
+      return $ case mbinders of
+        Just bndrs -> length bndrs > d
+        Nothing -> False
+
+-- | Given a name, check if that name is a type family. If
+-- so, return a list of its binders.
+#if MIN_VERSION_template_haskell(2,17,0)
+getTypeFamilyBinders :: Name -> Q (Maybe [TyVarBndr ()])
```
Comment (RyanGlScott) [inline src/Generics/Deriving/TH/Internal.hs:180]: You can avoid guarding this with CPP by using `TyVarBndr_ ()` instead, where [`TyVarBndr_`](https://hackage.haskell.org/package/th-abstraction-0.4.3.0/docs/Language-Haskell-TH-Datatype-TyVarBndr.html#t:TyVarBndr_) comes from `th-abstraction`.

**Comment by RyanGlScott** [comment]
Thanks! If I understand correctly, this is purely a refactoring with no change in user-visible behavior, right?

**Comment by treeowl** [comment]
Unless I've made a mistake, it should be mostly invisible. Might change a few error messages because of the `forall a -> ...` addition and such. The higher-rank error is improved by being thrown with `fail` rather than `error`.

**Comment by treeowl** [comment]
Oh, one last tiny change I want to make...

**Comment by treeowl** [comment]
OK, done.

**Review by RyanGlScott — APPROVED** [review]

**Pull request merged**